src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java

```diff
@@ -83,73 +90,42 @@ public void close() {
-        for (final DiscoveryNode node : clusterState.nodes()) {
+        for (final DiscoveryNode node : clusterChangedEvent.state().nodes()) {
             if (node.equals(localNode)) {
                 continue;
             }
             nodesToPublishTo.add(node);
         }
-        publish(clusterState, nodesToPublishTo, new AckClusterStatePublishResponseHandler(nodesToPublishTo, ackListener));
+        publish(clusterChangedEvent, nodesToPublishTo, new AckClusterStatePublishResponseHandler(nodesToPublishTo, ackListener));
     }
 
-    private void publish(final ClusterState clusterState, final Set<DiscoveryNode> nodesToPublishTo,
+    private void publish(final ClusterChangedEvent clusterChangedEvent, final Set<DiscoveryNode> nodesToPublishTo,
                          final BlockingClusterStatePublishResponseHandler publishResponseHandler) {
 
         Map<Version, BytesReference> serializedStates = Maps.newHashMap();
+        Map<Version, BytesReference> serializedDiffs = Maps.newHashMap();
 
         final AtomicBoolean timedOutWaitingForNodes = new AtomicBoolean(false);
         final TimeValue publishTimeout = discoverySettings.getPublishTimeout();
+        final boolean sendFullVersion = !discoverySettings.getPublishDiff() || clusterChangedEvent.previousState() == null;
+        final ClusterState clusterState = clusterChangedEvent.state();
+        Diff diff = null;
 
         for (final DiscoveryNode node : nodesToPublishTo) {
 
             // try and serialize the cluster state once (or per version), so we don't serialize it
             // per node when we send it over the wire, compress it while we are at it...
-            BytesReference bytes = serializedStates.get(node.version());
-            if (bytes == null) {
-                try {
-                    BytesStreamOutput bStream = new BytesStreamOutput();
-                    StreamOutput stream = CompressorFactory.defaultCompressor().streamOutput(bStream);
-                    stream.setVersion(node.version());
-                    ClusterState.Builder.writeTo(clusterState, stream);
-                    stream.close();
-                    bytes = bStream.bytes();
-                    serializedStates.put(node.version(), bytes);
-                } catch (Throwable e) {
-                    logger.warn("failed to serialize cluster_state before publishing it to node {}", e, node);
-                    publishResponseHandler.onFailure(node, e);
-                    continue;
+            boolean newlyAddedNode = !clusterChangedEvent.previousState().nodes().nodeExists(node.id());
+            if (sendFullVersion || newlyAddedNode) {
+                sendFullClusterState(clusterState, serializedStates, node, timedOutWaitingForNodes, publishTimeout, publishResponseHandler);
+            } else {
+                if (diff == null) {
+                    diff = clusterChangedEvent.state().diff(clusterChangedEvent.previousState());
```

Comment:
it would help me a lot if we could assign state() and previousState() to a variable instead of chaining all the methods